diff -Nur a/include/gbm.h b/include/gbm.h
--- a/include/gbm.h	2017-09-29 11:28:50.000000000 +0200
+++ b/include/gbm.h	2017-10-01 16:43:28.027784496 +0200
@@ -35,6 +35,7 @@
 
 #define __GBM__ 1
 
+#include <stddef.h>
 #include <stdint.h>
 
 /**
diff -Nur a/include/KHR/khrplatform.h b/include/KHR/khrplatform.h
--- a/include/KHR/khrplatform.h	2017-09-29 11:28:50.000000000 +0200
+++ b/include/KHR/khrplatform.h	2017-10-01 13:33:47.900316897 +0200
@@ -92,6 +92,8 @@
  *                                  int arg2) KHRONOS_APIATTRIBUTES;
  */
 
+#define GL_GLEXT_PROTOTYPES
+
 /*-------------------------------------------------------------------------
  * Definition of KHRONOS_APICALL
  *-------------------------------------------------------------------------
